### PR TITLE
WIP: Add ResolveScopeProvider extension 

### PR DESCRIPTION
**Pull request by arioston**
#480

### CODE REVIEW

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.phases.ExtensionPhase
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.ResolveScopeProvider
+import com.intellij.psi.search.GlobalSearchScope
+
+interface ResolveProviderSyntax {
+    fun IdeMetaPlugin.getResolveScope(
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt]:
```suggestion
    fun IdeMetaPlugin.resolveScope(
```

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                    ResolveScopeProvider.EP_NAME,
+                    object : ResolveScopeProvider(){
+                        override fun getResolveScope(file: VirtualFile, project: Project?): GlobalSearchScope? =
+                                getResolveScope(file, project)
+                    }
+            )
+
+    fun IdeMetaPlugin.globalSearchScope(
+        containsFile: (file: VirtualFile) -> Boolean,
+        isSearchInModuleContent: (aModule: Module) -> Boolean,
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt]:
```suggestion
        isSearchInModuleContent: (module: Module) -> Boolean,
```

Comment (i-walker) [reply idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt]: We should find a better name for `isSearchInModuleContent`

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    ): ExtensionPhase =
+            extensionProvider(
+                    ResolveScopeProvider.EP_NAME,
+                    object : ResolveScopeProvider(){
+                        override fun getResolveScope(file: VirtualFile, project: Project?): GlobalSearchScope? =
+                                getResolveScope(file, project)
+                    }
+            )
+
+    fun IdeMetaPlugin.globalSearchScope(
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:23]: nit: 
```suggestion
    fun ResolveProviderSyntax.globalSearchScope(
```
But feel free to merge :), whenever

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+     *
+     * IdeMetaPlugin().addResolveScopeProvider { file: VirtualFile, project: Project ->
+     *  if (file.fileType != KotlinFileType.INSTANCE) return@addResolveScopeProvider null
+     *
+     *  val ktFile = PsiManager.getInstance(project).findFile(file) as? KtFile ?: return@addResolveScopeProvider null
+     *  val scriptDefinition = ktFile.findScriptDefinition()
+     *  when {
+     *      scriptDefinition == null -> null
+     *      scriptDefinition.baseClassType.fromClass == Any::class -> null
+     *      scriptDefinition.asLegacyOrNull<StandardIdeScriptDefinition>() != null -> null
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:44]: the null cases can be removed for redudancy

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+     *      scriptDefinition is ScriptDefinition.FromConfigurations ||
+     *              scriptDefinition.asLegacyOrNull<KotlinScriptDefinitionFromAnnotatedTemplate>() != null -> {
+     *           GlobalSearchScope.fileScope(project, file)
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:47]: A clarification what this match does in combination with the ResolveScopeProvider would be awesome

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+     * import org.jetbrains.kotlin.idea.KotlinFileType
+     * import org.jetbrains.kotlin.idea.core.script.ScriptConfigurationManager
+     * import org.jetbrains.kotlin.idea.core.script.StandardIdeScriptDefinition
+     * import org.jetbrains.kotlin.psi.KtFile
+     * import org.jetbrains.kotlin.scripting.definitions.ScriptDefinition
+     * import org.jetbrains.kotlin.scripting.definitions.findScriptDefinition
+     * import org.jetbrains.kotlin.scripting.resolve.KotlinScriptDefinitionFromAnnotatedTemplate
+     *
+     * IdeMetaPlugin().addResolveScopeProvider { file: VirtualFile, project: Project ->
+     *  if (file.fileType != KotlinFileType.INSTANCE) return@addResolveScopeProvider null
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:37]: Do something with `file.takeIf {it.fileType}?.let { ...}`.

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ */
+interface ResolveProviderSyntax {
+
+
+    /**
+     * registers an [GlobalSearchScope].
+     * One minimal example from [KotlinScriptResolveScopeProvider], may look like this:
+     * ```kotlin
+     * import arrow.meta.ide.IdeMetaPlugin
+     * import arrow.meta.Plugin
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:23]: This PR uses the old encoding. Master has now `CliPlugin` and `IdePlugin`

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt (hunk 6 above)
Comment (arioston) [reply idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:37]: something like this:


```
import arrow.meta.ide.IdeMetaPlugin
import arrow.meta.phases.ExtensionPhase
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiManager
import com.intellij.psi.search.GlobalSearchScope
import org.jetbrains.kotlin.idea.KotlinFileType
import org.jetbrains.kotlin.idea.core.script.ScriptConfigurationManager
import org.jetbrains.kotlin.scripting.definitions.ScriptDefinition
import org.jetbrains.kotlin.scripting.definitions.findScriptDefinition
import org.jetbrains.kotlin.scripting.resolve.KotlinScriptDefinitionFromAnnotatedTemplate

val IdeMetaPlugin.scopeProvider: ExtensionPhase
    get() = addResolveScopeProvider { file: VirtualFile, project: Project ->
        file.takeIf { it.fileType == KotlinFileType.INSTANCE }?.let {
            PsiManager.getInstance(project).findFile(it)?.findScriptDefinition()
                    ?.let { scriptDefinition ->
                        when {
                            scriptDefinition is ScriptDefinition.FromConfigurations || scriptDefinition.asLegacyOrNull<KotlinScriptDefinitionFromAnnotatedTemplate>() != null -> {
                                GlobalSearchScope.fileScope(project, file)
                                        .union(ScriptConfigurationManager.getInstance(project).getScriptDependenciesClassFilesScope(file))
                            }
                            else -> null
                        }

                    }
        }
    }
```

Comment (i-walker) [reply idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:37]: That looks better. The last `when` is not needed. 
Something like:
```kotlin
file.takeIf {
  it.fileType == KotlinFileType.INSTANCE && 
  PsiManager.getInstance(project)
   .findFile(it)
   ?.findScriptDefinition()
   ?.safeAs<ScriptDefinition.FromConfigurations>()
   ?.asLegacyOrNull<KotlinScriptDefinitionFromAnnotatedTemplate>()
      != null
  }.let { file ->
   GlobalSearchScope
    .fileScope(project, file)
    .union(
      ScriptConfigurationManager
      .getInstance(project)
      .getScriptDependenciesClassFilesScope(file)
     )
}
```

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import arrow.meta.phases.ExtensionPhase
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.ResolveScopeProvider
+import com.intellij.psi.search.GlobalSearchScope
+
+/**
+ * [GlobalSearchScope] search for {@code VirtualFile}s in global scope.
+ * [ResolveProviderSyntax] provides APIs to create GlobalSearchScope.
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:13]:
```suggestion
```
Let's remove that part and move it down.

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+ */
+interface ResolveProviderSyntax {
+
+
+    /**
+     * registers an [GlobalSearchScope].
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:19]:
```suggestion
     * registers an ResolveScopeProvider.
     * [GlobalSearchScope] defines a Scope based on [VirtualFile]s and Modules.
```

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+     *                  .union(
+     *                          ScriptConfigurationManager
+     *                                 .getInstance(project)
+     *                                 .getScriptDependenciesClassFilesScope(file)
+     *                  )
+     *  }
+     * }
+     *
+     *
+     * @param getResolveScope defines where the [VirtualFile] is in the [Project] return the [GlobalSearchScope] if there is one.
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:56]:
```suggestion
```
That adds confusion IMO

Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt:56]:
```suggestion
```
That adds confusion IMO

**Pull request merged**